cpp/src/arrow/io/buffered_test.cc

```diff
@@ -472,6 +480,24 @@ TEST_F(TestBufferedInputStream, SetBufferSize) {
   ASSERT_OK(buffered_->SetBufferSize(5));
 }
 
+// GH-43060: Internal buffer should not greater than the
+// bytes could buffer.
+TEST_F(TestBufferedInputStream, BufferSizeLimit) {
+  {
+    MakeExample1(/*buffer_size=*/100000, default_memory_pool(),
+                 /*fill_raw_read_bound=*/true);
+    // buffer_sizes should be limited to the size of the data
+    EXPECT_EQ(test_data_.size(), buffered_->buffer_size());
+  }
+
+  {
+    MakeExample1(/*buffer_size=*/10, default_memory_pool(), /*fill_raw_read_bound=*/true);
+    ASSERT_OK(buffered_->Read(10));
+    ASSERT_OK(buffered_->SetBufferSize(/*new_buffer_size=*/100000));
+    EXPECT_EQ(test_data_.size() - 10, buffered_->buffer_size());
+  }
```

Comment:
Ideally this would be something like:
```suggestion
  {
    MakeExample1(/*buffer_size=*/10, default_memory_pool(), /*raw_read_bound=*/15;
    ASSERT_OK(buffered_->Read(10));
    ASSERT_OK(buffered_->SetBufferSize(/*new_buffer_size=*/100000));
    EXPECT_EQ(5, buffered_->buffer_size());
  }
```